draft-ietf-dprive-dnsoquic.md

```diff
@@ -394,12 +394,17 @@ A client MAY take advantage of the connection resume mechanisms supported by
 QUIC transport {{!RFC9000}} and QUIC TLS {{!RFC9001}}. Clients SHOULD consider
 potential privacy issues associated with session resume before deciding to use
 this mechanism. These privacy issues are detailed in
-{{privacy-issues-with-session-resume}}.
+{{privacy-issues-with-session-resume}} and {{privacy-issues-with-0-rtt-data}},
+and the implementation considerations are discussed in
+{{using-0-rtt-and-resumption}}.
 
-When resuming a session, a client MAY take advantage of the 0-RTT mechanism
-supported by QUIC. The 0-RTT mechanism MUST NOT be used to send data that is
+The 0-RTT mechanism MUST NOT be used to send data that is
 not "replayable" transactions. For example, a client MAY transmit a Query as
-0-RTT, but MUST NOT transmit an Update.
+0-RTT, but MUST NOT transmit an Update. Servers that receive requests
+for replayable transactions MUST NOT process them before the connection
+handshake is confirmed, as defined in section 4.1.2 of {{!RFC9001}}; 
+servers MAY close connections in which replayable transactions are
+attempted with the error code DOQ_PROTOCOL_ERROR.
```

Comment:
Yes, makes sense. There are three plausible behavior, in order of desirability:

* Queue the request, and only process it after the connection is confirmed
* Reject the request, with an error of type "TOO_EARLY" (TBD). probably with RESET_STREAM, because the client has already sent the whole stream and set the FIN bit.
* Close the connection with either "PROTOCOL_ERROR" or "TOO_EARLY".

This is DoQ specific, so it make sense to process that at the DoQ layer, rather than DNS. But Sara may have another opinion.

Reply:
On the other hand, I like to put the onus on the client on this. If we keep the text saying "clients must not [do bad stuff]", then doing such stuff is a protocol error, and servers MAY just close the connection and yell. If on the contrary we say that servers should queue the offending request until the session is confirmed, then that means that the clients may just as well push these requests in 0-RTT.

So the question is, do we see a compelling advantage for clients pushing the request in 0-RTT, versus waiting for connection complete and pushing them? Yes, the requests will be ready immediately after the server receives the "client finished" signal, by opposition to "after the first 1-RTT packet from the client is received". Those two events are likely happening at the same time. Is that worth the extra complexity?

Reply:
It's a shared risk.  In HTTP, we have some expectations about what is safe that are generic (GET is good) and some that are specific to the request (some POST requests are perfectly OK, but not all of them).  Sometimes the client will know that it is safe, sometimes they will think it is safe and learn that it is not.  This latter one is the key risk: if the server knows that a query is unsafe (because of how it manages the cache for that particular entry, or other reasons), then it needs to be able to reject things, likely on a per-query basis.

For DNS, I think that enqueuing queries is probably unwise in most cases.  A TOO_EARLY or similar response at some layer is probably fine, though I would tend toward using DNS-layer signaling.  That is for the same reason we use HTTP status codes in RFC 8470: it makes the information available to the DNS client proper, not just the adapter that is enabling DoQ, which might only be an intermediary.

Reply:
If this were to be failed at the DNS layer, then the server would have to send REFUSED and we would need to define a transport specific EDE code to signal TOO_EARLY and then require all implementation of DoQ to implement EDE, which does feel odd to me. 
I think this question should go to the list for discussion because it is a functional change to the spec.

Reply:
The server can send a refuse response for any query, not just those defined in 0-RTT. This does not require definition of a QUIC-level error. I think reusing the REFUSED error is simpler than adding a transport-level mechanism.

Reply:
Yeah, I think that REFUSED + EDE(TOO_EARLY) sounds like a good plan.  I agree that list discussion is appropriate.